go/teams/fast.go

```diff
@@ -72,43 +95,945 @@ func (f *FastTeamChainLoader) verifyTeamNameViaParentLoad(m libkb.MetaContext, i
+			PerTeamKeySeedsVerified: make(map[keybase1.PerTeamKeyGeneration]keybase1.PerTeamKeySeed),
+			DownPointers:            make(map[keybase1.Seqno]keybase1.DownPointer),
+			LinkIDs:                 make(map[keybase1.Seqno]keybase1.LinkID),
+		},
+	}
+}
+
+// refresh the team's state, but loading with the server. It will download new stubbed chainlinks,
+// fill in unstubbed chainlinks, make sure that prev pointers match, make sure that the merkle
+// tree agrees with the chain tail, and then run the audit mechanism.
+func (f *FastTeamChainLoader) refresh(m libkb.MetaContext, arg fastLoadArg, state *keybase1.FastTeamData, shoppingList shoppingList) (res *keybase1.FastTeamData, err error) {
+
+	defer m.CTrace(fmt.Sprintf("FastTeamChainLoader#refresh(%+v)", arg), func() error { return err })()
+
+	groceries, err := f.loadFromServerWithRetries(m, arg, shoppingList)
+	if err != nil {
+		return nil, err
+	}
+
+	// Either makes a new state, or deepcopies the existing state, so that in the case
+	// of an error, we haven't corrupted what's in cache. Thus, from here on out,
+	// we are playing with our own (unshared) copy of the state.
+	state = makeState(arg, state)
+
+	// check that all chain links sent down form a valid hash chain, and point
+	// to what we already in had in cache.
+	err = f.checkPrevs(m, state.Chain.Last, groceries.newLinks)
+	if err != nil {
+		return nil, err
+	}
+
+	// check that the server stubbed properly.
+	err = f.checkStubs(m, shoppingList, groceries.newLinks)
+	if err != nil {
+		return nil, err
+	}
+
+	// peform a probabilistic audit on the new links
+	err = f.audit(m, arg.ID, arg.Public, groceries.newLinks)
+	if err != nil {
+		return nil, err
+	}
+
+	err = f.mutateState(m, arg, state, groceries)
+	if err != nil {
+		return nil, err
+	}
+
+	return state, nil
+}
+
+// updateCache puts the new version of the state into the cache on the team's ID.
+func (f *FastTeamChainLoader) updateCache(m libkb.MetaContext, state *keybase1.FastTeamData) {
+	f.getLRU().Add(state.Chain.ID, state)
+}
+
+// loadLocked is the inner loop for loading team. Should be called when holding the lock
+// this teamID.
+func (f *FastTeamChainLoader) loadLocked(m libkb.MetaContext, arg fastLoadArg) (res *fastLoadRes, err error) {
+
+	state := f.findStateInCache(m, arg)
+
+	var shoppingList shoppingList
+	if state != nil {
+		shoppingList.computeWithPreviousState(m, arg, state)
+		if shoppingList.isEmpty() {
+			return f.toResult(m, arg, state)
+		}
+	} else {
+		shoppingList.computeFreshLoad(m, arg)
+	}
+
+	m.CDebugf("FastTeamChainLoader#loadLocked: computed shopping list: %+v", shoppingList)
+
+	state, err = f.refresh(m, arg, state, shoppingList)
+	if err != nil {
+		return nil, err
+	}
+	f.updateCache(m, state)
+
+	return f.toResult(m, arg, state)
+}
+
+// newLRU installs a new LRU for the loader and purges the old one. Does a swap to avoid race conditions
+// around logging out.
```

Comment:
Why swap? The lru is threadsafe. Wouldn't it be simpler to initialize the lru in the constructor, Purge here, and do away with the `lruMutex`?

Reply:
what about: (1) start load; (2) user logs out and logs back in; (3) finish load.  now user B has user A's cache entry.

Reply:
How do the strategies differ when it comes to that problem?

Reply:
in my strategy, you're putting the cache to the old LRU, which is going to be garbage collected since nothing points to it.

Reply:
`updateCache` doesn't take the old LRU as a parameter, it calls `getLRU()`. If the LRU was only `getLRU()`d once per Load call then I'd see what you're saying. Then that race would still return the team which is bad, but not store it which is good.

Reply:
ok, i see.  i can fix that.

Reply:
ok, fixed